### ISSUE TITLE
First client map set often takes 5s after connect resulting connect/reconnect loop

### ISSUE THREAD

**Issue by jtorkkel** [hazelcast/hazelcast#17887]
**Background**
We use Hazelcast 4.01 in client/server mode in docker environment. We have fixed number of members in cluster, and use fixed DNS names between members. 
Client is in smart routing mode and use list of DNS addresses to connect to cluster.

We have noticed two issues
1. When cluster member IP-address change the client connect keep failing forever because it seems to use old member IP-address. We have address this by reloading the config/reconnect on error/timeout, this workaround is probably needed due reported 4.1.1 open bug to address cluster address change: https://github.com/hazelcast/hazelcast/issues/17062
2. After connect the first map set sometimes takes 5s. We have implemented timeout to address case 1 issue, which result disconnect, reconfigure and reconnect and again when first write take 5s. This issue might follow when cluster is reconfigured but might also happen randomly as we have seen random set/get response times up to 10s, which trigger reconnect and first et always takes 5s. Thus connection might end up into almost infinite the connect/reconnect loop unless the first set is shorter than timeout 5s or long response timeout. In most cases the reconnect look is infinite and restart is required.

We have used different timeout settings like
- 1s --> reconnect loop
- 2s --> reconnect loop
- 4s --> reconnect loop
- 10s newer triggering, first reconnect address the case 1) IP-address issue but we end into reconnect loop as set always takes 5s.

**Problem description**
Sounds that there is pseudorandom fixed 5s initial delay for set which trigger our custom timeout handler but no fixed timeout value can fix he issue because < 5s timeout result reconnect loop and > 5s timeout is never triggering or triggering very seldom. The client events cannot be trusted as CONNECT event often missing based on logs after successful connect. 
reconnect loop might be due dynamic runtime reconfiguration/reconnect or too frequent reconnect because restarting the service which takes 1min fix the issues.

We can possible try to workaround this by preventing to reconnect during first 10s (or after 60s depending what is causing the issue) after connect as example but there is no guarantee that issue can be fixed.
Fixing all the issues on top of Hazelcast client library is not desirable because having long timeout cause robustness issues as Hazelcast clients do not want to have +1s timeouts. Long timeouts cause issues when cluster is repartitioning and better to have fast timeout and fall back to no cache during those period instead of driver having custom login with long timeout to try to recover.

**Expected behavior:**
- Hazelcast client can connect reliable to cluster
- Hazelcast client can automatically trigger reconnect when needed
- Hazelcast client implement reliable timeout or proper error code to client if reconnect is needed

Not known if related to
- Reduce delays when joining a cluster https://github.com/hazelcast/hazelcast/pull/17428
- Hazelcast node connect performance slow takes ~10s https://github.com/payara/Payara/issues/4858 
- Fix excessive delays when joining a cluster (5-10 seconds minimum) https://github.com/hazelcast/hazelcast/issues/17427
but probably different issue as some client works poperly and these are no cluster partitioning ongoing

**To Reproduce**
- Based on metrics every time the response time exceed the timeout and or custom timeout trigger reconnect there is danger for reconnect loop
- Reconfiguring cluster also trigger reconnect logic multiple times

Common details that we're often interested in:
- Cluster size 6
- Number of the clients, 4
- Version of Java. 8.
- Operating system. Linux

**Comment by sancar** [comment]
Hi @jtorkkel , The first issue is fixed and will be released in 4.1.1. Fix -> https://github.com/hazelcast/hazelcast/pull/17911

> We have address this by reloading the config/reconnect on error/timeout

Can you be more specific about this sentence? What do you mean by reloading the config? 
Do you reopen a client with `HazelcastClient.newHazelcastClient()`  ?


> After connect the first map set sometimes takes 5s
Is the cluster in the same LAN, if yes 5s is too long for that? Or are you connecting to a remote cluster over the internet?


> The client events cannot be trusted as CONNECT event often missing based on logs after successful connect.

This should never happen. Note that CONNECT means that client is connected to at least one member in the cluster. 
Can you share how you start the client with the exact configuration and also the logs of the client?

**Comment by jtorkkel** [comment]
Q: Do you reopen a client with HazelcastClient.newHazelcastClient() ?
A: Yes. 

Q: Is the cluster in the same LAN, if yes 5s is too long for that? Or are you connecting to a remote cluster over the internet?
A: 3 members are in same cluster, 3 in other DC but connected with fiber, in separate partition group. During normal operation DC delay <<1ms, naturally there might be random longer response times. 
When issue happen all SETS takes exactly 5s while all GET are immediate. Not possible that each and every SET takes 5s, there is artificial persisting delay in place. Only some clients are impacted with 5s delay, some other clients have fast SET while some has fixed 5s. 

When SET takes too long time we have timeout and we shutdown the client. During shutdown some SET result exception:
Caused by: com.hazelcast.client.HazelcastClientNotActiveException: Client is not active.
Once we do not get exceptions the SET apparently succeed but takes 5s, if our timeout is shorter than 5s we disconnect and reconnect, and first SET timeout again.

Q: Can you share how you start the client with the exact configuration and also the logs of the client?

```
 private static ClientConfig clientConfig(HzCacheClientProperties props, int rev) {
        ClientConfig config = new ClientConfig()
                .setProperty("hazelcast.logging.type", "slf4j")
                .setProperty("hazelcast.phone.home.enabled", "false")
                .setProperty("hazelcast.client.heartbeat.timeout", String.valueOf(props.getHeartBeatTimeout().toMillis()))         // default 60000
                .setProperty("hazelcast.client.invocation.timeout.seconds", String.valueOf(props.getInvocationTimeout().getSeconds()))  // default 1000
                .setProperty("hazelcast.client.statistics.enabled", String.valueOf(props.isStatisticsEnabled()))
                .setProperty("hazelcast.diagnostics.enabled", String.valueOf(props.isDiagnosticsEnabled()))
                .setClusterName(props.getClusterName().replace(".", "_"))
                .setInstanceName(props.getClientName() + "-r" + rev);

        config
                .getNetworkConfig()
                .setAddresses(props.getMembers())
                .setSmartRouting(true)
                .setConnectionTimeout((int) props.getConnectionTimeout().toMillis()); // timeout, 0 until success

        HzCacheClientProperties.ConnectionRetry retry = props.getRetry();
        config
                .getConnectionStrategyConfig()
                .setAsyncStart(true)
                .getConnectionRetryConfig()
                .setInitialBackoffMillis((int) retry.getInitialBackoff().toMillis())
                .setMaxBackoffMillis((int) retry.getMaxBackoff().toMillis())
                .setMultiplier(retry.getMultiplier())
                .setClusterConnectTimeoutMillis(retry.getConnectTimeout().toMillis())  // to trigger client reset
                .setJitter(retry.getJitter());

        log.info("HzClient config {}", config);
        return config;
    }
```

**Comment by mdumandag** [comment]
Can this be related to this issue? https://github.com/hazelcast/hazelcast/issues/9398

The 5-second delay might come from the `hazelcast.client.operation.backup.timeout.millis` property, in case of missing backups for set operations

**Comment by jtorkkel** [comment]
Could be, behaves similar way.
However there are no errors on cluster logs. Typically 2 instance works and 2 has 5s SET delay. The failing instance use two different map, both has 5s delay. At the same few other service which use different map name works without any problems. This hints that issue is either map specific or actually client specific. Restarting or reconnecting the client fix the issue but require complex logic (due reactive model we use) to guess when to reconnect and when not to eliminate timeouts immediately after connect. If backup-ack is the issue why the ack is permanently missing, normally 5s delay continues few hours (I think it might recover) or until client is restarted. 

Could it be that restarting one cluster member (docker-->meber address change) trigger chain reaction which result the lost of ACK when connected to cluster during re-partitioning ongoing. I assume that #17911 will not result reconnect in that case, #17911 fix only connect issue and client needs to reconnect in case 5s timeout. 

```
When an operation with sync backup is sent by a client to the Hazelcast member(s), the acknowledgment of the operation’s backup is sent to the client by the backup replica member(s). This improves the performance of the client operations.
By default, backup acknowledgement to the client is enabled for smart clients (unisocket clients do not support it). 
```

Would it help disabling backup-ack, or would there be high performance penalty?
Or would async backup help ?

I assume with async backup or or without ack there is danger that data is lost if one member is restarted.
Most map we use are just cache, some are used to store state. 
Due the fact that client in client-server setup client cannot configure new maps configuration (new feature request) we use same default map config (sync, backup-count-1, TTL 300s) for every map regardless if created map is used for cache or to store the state. 
Having sync map for state and async for cache would complicate the config because for each new map holding state we should reconfigure the cluster/server with corresponding map name config (sync) to eliminate the default map config usage (which would be async).

No easy workaround on client side.
```
Hazelcast supports both synchronous and asynchronous backups. By default, backup operations are synchronous and configured with backup-count. In this case, backup operations block operations until backups are successfully copied to backup members (or deleted from backup members in case of remove) and acknowledgements are received. Therefore, backups are updated before a write(put, set, remove and their async counterparts) operation is completed, provided that the cluster is stable. Sync backup operations have a blocking cost which may lead to latency issues.

Asynchronous backups, on the other hand, do not block operations. They are fire & forget and do not require acknowledgements; the backup operations are performed at some point in time.
```

**Comment by sancar** [comment]
About `CLIENT_CONNECTED` of the client. I see that you are starting in the `async` mode. 
`.setAsyncStart(true)`
You need to attach your listener before the client starts so that you can deterministically catch the event always. 
Since it is absent from the config, I assume that you are configuring that after you open the client. More specifically, 
`HazelcastClient.newHazelcastClient()` 

You can register your lifecycle listener in the configuration as follows:
```
       ClientConfig clientConfig = new ClientConfig();
        clientConfig.addListenerConfig(new ListenerConfig().setImplementation(new LifecycleListener() {
            @Override
            public void stateChanged(LifecycleEvent event) {

            }
        }));

        HazelcastClient.newHazelcastClient(clientConfig);
```

> Would it help disabling backup-ack, or would there be high performance penalty?
Yes, there will be a penalty. How high the penalty will heavily depend on the setup you have. I recommend you to measure first to decide. 
Having said that, we are not sure that is the problem yet. We can use disabling-backup acks to pinpoint the problem. If it continues even if we disable it, we can be sure that it is not the cause. But if we can't reproduce the problem, we will not be sure. 
Please set the config below and try:
```
clientConfig.setBackupAckToClientEnabled(false);
```

Would it be possible to send logs of the client in the case of failure for us to investigate further, preferably with debug logs are enabled?

**Comment by jtorkkel** [comment]
We set up the listener immediately after the config, but will make the propose change to set it as part of the config.

We have seen the issue only in production when we had 10s timeout, with 2s or 4s timeout cannot be seen as we disconnect before 5s timeout. 

Seems that our reconnect logic is broken during cluster member restart / long network delays while high load. High enough load cause that that "multiple threads" have pending request, and when we disconnect/reconnect due 2s timeout the other "threads" cannot be canceled. This result that some other threads might timeout immediately after connect completed resulting new reconnect (due lack of guard period after reconnect). Probably request which are done before timeout cause the issue and re-configuraton/reconnect result they never get reply nor exception, whereas the request which are done during the connect result Client Not active WARN message. And making few disconnect/reconnect just request more queuing. We do not want to bring health down because that would cause even bigger break.

We will retest with 10s timeout and debug logs to see if 5s issue can be reproduced.

It is clear that we need to add guard/filtering to or reconnect logic, as reconnect logic fix cluster member address change but cause other issues. 
Maybe once 4.1.1 with #17911 is out we can remove the reconfig logic and rely that native Hazelcast client can connect and reconnect reliable. And maybe tune the reconnect/reconfig logic to trigger only if error situation persists.

**Comment by sancar** [comment]
Hi @jtorkkel, The issue seems to be specific to your setup. I would like to close the issue for now. When you have applied the fixes with 4.1.1 if you have problems feel free to open a new issue.

**Comment by jtorkkel** [comment]
Hi,

We can now repro this 100% simply by restaring one Hazelcast cluster member in docker which result IP-address change, but no change in member name. Cluster will detect the member is lost, then start migration from 6-->5 members, after migration new member if detected and new migration performed 5->6.

Based on logs clients are not reconnected during neither of the migration.


We also set 
```
configuration.setProperty("hazelcast.client.operation.backup.timeout.millis", "7000");
```

client configuration
```
xxx.hazelcast.members=v1-0:5701,v1-1:5701,v1-2:5701,v1-3:5701,v1-4:5701,v1-5:5701

```
cluster configuration:
```
hazelcast.port=5701
hazelcast.members=v1-0:5701,v1-1:5701,v1-2:5701,v1-3:5701,v1-4:5701,v1-5:5701
hazelcast.partition-group-a=v1-0,v1-2,v1-4
hazelcast.partition-group-b=v1-1,v1-3,v1-5
```

And now set time is **7000ms** for ~30% of the sets.

Seems that hazelcast smart client does not get correct partition update from cluster after the migration and thus some set, probably to the IP-address of the lost members get timeout.
